hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerStateManager.java

```diff
@@ -325,8 +325,12 @@ public void addContainerInfo(long containerID,
                                Pipeline pipeline) throws IOException {
     Preconditions.checkNotNull(containerInfo);
     containers.addContainer(containerInfo);
-    pipelineManager.addContainerToPipeline(pipeline.getId(),
-        ContainerID.valueof(containerID));
+    if (pipeline != null) {
```

Comment:
I'm trying to understand why this is not an issue in SCM? 
Does Recon have similar idea like SCM safemode to ensure Pipeline meet the minimal requirement?

Reply:
@xiaoyuyao Good question. Although we get pipeline report for a new pipeline through DN, it does not contain all the information that was used by SCM to create the pipeline (Factor, other nodes). While implementing this, we considered 2 ways to handle this, either through an elaborate pipeline creation ACK command which Recon can pick up, or through Recon requesting SCM for a confirmation whenever it sees a new pipeline. We went with the latter.  

In general, any state change that "originates" in the SCM needs to be synced with Recon. But anything that originates in the Datanode works automatically.